R/expected_event.R

```diff
@@ -135,50 +135,42 @@ expected_event <- function(
   # Divide the time line into sub-intervals ----
 
   ## by piecewise enrollment rates
-  df_1 <- data.frame(start_enroll = c(0, cumsum(enroll_rate$duration)))
-  df_1$end_fail <- total_duration - df_1$start_enroll
+  start_enroll <- c(0, cumsum(enroll_rate$duration))
+  df_1 <- data.frame(start_enroll, end_fail = total_duration - start_enroll)
   df_1 <- df_1[df_1$end_fail > 0, ]
 
   ## by piecewise failure & dropout rates
+  end_fail <- cumsum(fail_rate$duration)
   df_2 <- data.frame(
-    end_fail = cumsum(fail_rate$duration),
+    end_fail = end_fail,
     fail_rate_var = fail_rate$fail_rate,
-    dropout_rate_var = fail_rate$dropout_rate
+    dropout_rate_var = fail_rate$dropout_rate,
+    start_enroll = total_duration - end_fail
   )
-  df_2$start_enroll <- total_duration - df_2$end_fail
 
-  temp <- cumsum(fail_rate$duration)
-  if (temp[length(temp)] < total_duration) {
-    df_2 <- df_2[-nrow(df_2), ]
-  } else {
-    df_2 <- df_2[df_2$start_enroll > 0, ]
-  }
+  N <- length(end_fail)
+  df_2 <- df_2[if (end_fail[N] < total_duration) -N else end_fail < total_duration, ]
```

Comment:
@LittleBeannie This was the line that I mentioned during the meeting---I was wondering why we'd want to remove the last row on the condition `if (end_fail[N] < total_duration)`.

Reply:
Hi @yihui , thanks for sharing your simplification. After researching some special cases, your code looks okay, but I personally prefer to keep it as it is because the calculation of expected events is very complicated (source: https://merck.github.io/gsDesign2/articles/story-compute-expected-events.html). I am concerned that a one-line change in the code could impact cases not covered in the tests.

Additionally, here are some minor coding practices we previously followed in gsDesign2:
- Use lowercase for all variable names.
- The letter `N` is usually reserved for sample size, so please replace it with other letters if you prefer!

Reply:
@LittleBeannie I meant it feels to me that this line could have been simplified to:

```suggestion
  df_2 <- df_2[end_fail < total_duration, ]
```

and I was asking why we needed the `if` here. The current code reads "if the last cumsum of duration is smaller than total duration, keep all rows except the last row, otherwise keep the rows where the cumsum of duration is smaller than total duration". That logic doesn't sound right to me. In the case of `end_fail[N] < total_duration`, I feel we don't need to filter `df_2`.

BTW, as I mentioned in the meeting, our tests would still pass if I remove the `if` here.

(I'll revise the coding style at the end---thanks for the reminder!)